### PR TITLE
docment GAN

### PR DESCRIPTION
**Pull request by tmabraham**
Okay here's my first docment PR, let me know if there are any issues and I would appreciate any feedback you have.

I'll note I did take out the `after_epoch` and `show_img` arguments since they weren't doing anything. I will take a little time to debug and get them working. Shall I make an issue in the main fastai repo? I can work on fixing that after the docment sprint.

### CODE REVIEW

**Comment by tmabraham** [comment]
This PR would finish https://github.com/muellerzr/fastai-docment-sprint/issues/20.

File: fastai/vision/gan.py [hunk 1]
```diff
@@ -46,7 +58,13 @@ def forward(self, x): return x.view(*(list(x.shape)+[1]*self.n_dim))
...
 # Cell
 @delegates(ConvLayer.__init__)
-def basic_generator(out_size, n_channels, in_sz=100, n_features=64, n_extra_layers=0, **kwargs):
+def basic_generator(out_size, # output size for the generator (same as the input size for the critic)
+                    n_channels, # number of channels of the output of the generator
+                    in_sz=100, # size of the input noise vector for the generator
+                    n_features=64, # number of features used in the generator
+                    n_extra_layers=0, # number of extra hidden layers in the generator
+                    **kwargs
+                    ) -> nn.Sequential:
```
Comment (warner-benjamin) [inline fastai/vision/gan.py:67]: Most of the documented methods have some [style](https://github.com/muellerzr/fastai-docment-sprint/blob/master/docs_src/style-guide.md) issues. This method was missing type hints, missing the newline if not `self` or `cls`, missing capitalization, and had too many tabs for args.

If you could fix these formatting issues in other methods, everything looks pretty good.
```suggestion
def basic_generator(
    out_size:int, # Output size for the generator (same as the input size for the critic)
    n_channels:int, # Number of channels of the output of the generator
    in_sz:int=100, # Size of the input noise vector for the generator
    n_features:int=64, # Number of features used in the generator
    n_extra_layers:int=0, # Number of extra hidden layers in the generator
    **kwargs
) -> nn.Sequential:
```
One thing I am unsure of is whether we are adding docments to kwargs.

**Review by warner-benjamin — CHANGES_REQUESTED** [review]
If you could fix the highlighted formatting issues from `basic_generator` in other methods, everything looks pretty good.

File: fastai/vision/gan.py (hunk 1 above)
Comment (warner-benjamin) [reply fastai/vision/gan.py:67]: Update. We are not adding docments to kwargs, relying on @delegates where applicable

File: fastai/vision/gan.py [hunk 2]
```diff
@@ -12,21 +12,34 @@
 # Cell
 class GANModule(Module):
     "Wrapper around a `generator` and a `critic` to create a GAN."
-    def __init__(self, generator=None, critic=None, gen_mode=False):
+    def __init__(self,
+        generator:nn.Module=None, # The generator PyTorch module
+        critic:nn.Module=None,  # The discriminator PyTorch module
+        gen_mode:(None,bool)=False # Whether or not the GAN should be set to generator mode
+    ):
```
Comment (warner-benjamin) [inline fastai/vision/gan.py:19]:
```suggestion
    def __init__(self,
        generator:nn.Module=None, # The generator PyTorch module
        critic:nn.Module=None, # The discriminator PyTorch module
        gen_mode:(None,bool)=False # Whether the GAN should be set to generator mode
    ):
```

File: fastai/vision/gan.py [hunk 3]
```diff
@@ -276,7 +338,11 @@ def show_results(x:InvisibleTensor, y:TensorImage, samples, outs, ctxs=None, max
     return ctxs
 
 # Cell
-def gan_loss_from_func(loss_gen, loss_crit, weights_gen=None):
+def gan_loss_from_func(
+    loss_gen:callable, # A loss function for the generator. It should take in the generator output images and target real images and provide an evaluation.
+    loss_crit:callable, # A loss function for the critic. It should take in the images, real or fake, and also labels indicating if it's real or fake, and provide an evaluation.
+    weights_gen:(None, list, tuple)=None # Weights for the generator and critic loss function
+):
```
Comment (warner-benjamin) [inline fastai/vision/gan.py:345]: Isn't the `loss_crit` only passed labels and not any images?

Also rephrased to be more concise.
```suggestion
def gan_loss_from_func(
    loss_gen:callable, # A loss function for the generator. Evaluates generator output images and target real images
    loss_crit:callable, # A loss function for the critic. Evaluates predictions of real and fake images.
    weights_gen:(None, list, tuple)=None # Weights for the generator and critic loss function
):
```

File: fastai/vision/gan.py [hunk 4]
```diff
@@ -298,8 +364,21 @@ def _tk_diff(real_pred, fake_pred): return real_pred.mean() - fake_pred.mean()
...
     "A `Learner` suitable for GANs."
-    def __init__(self, dls, generator, critic, gen_loss_func, crit_loss_func, switcher=None, gen_first=False,
-                 switch_eval=True, show_img=True, clip=None, cbs=None, metrics=None, **kwargs):
+    def __init__(self,
+        dls:DataLoaders, # DataLoaders object for GAN data
+        generator:nn.Module, # Generator model
+        critic:nn.Module, # Critic model
+        gen_loss_func:callable, # Generator loss function
+        crit_loss_func:callable, # Critic loss function
+        switcher:(Callback,None)=None, # Callback for switching between generator and critic training (`FixedGANSwitcher` is used if set to `None`)
```
Comment (warner-benjamin) [inline fastai/vision/gan.py:373]:
```suggestion
        switcher:(Callback,None)=None, # Callback for switching between generator and critic training, defaults to `FixedGANSwitcher`
```

**Review by warner-benjamin — CHANGES_REQUESTED** [review]
Noticed one formatting error. Made a couple suggestions to make the doc strings more concise. Please replicate those in the other doc strings. Also, I think the `loss_crit` doc string had an error, but am not 100% sure.

File: fastai/vision/gan.py (hunk 3 above)
Comment (tmabraham) [reply fastai/vision/gan.py:345]: You are right, thanks for the correction!

**Comment by tmabraham** [comment]
@warner-benjamin updated, let me know if there's anything else that needs to be updated...

File: fastai/vision/gan.py [hunk 5]
```diff
@@ -309,14 +388,28 @@ def __init__(self, dls, generator, critic, gen_loss_func, crit_loss_func, switch
...
 
     @classmethod
-    def from_learners(cls, gen_learn, crit_learn, switcher=None, weights_gen=None, **kwargs):
+    def from_learners(cls,
+        gen_learn:Learner, # A `Learner` object that has the generator
+        crit_learn:Learner, # A `Learner` object that has the critic
+        switcher:(Callback,None)=None, # Callback for switching between generator and critic training, defaults to `FixedGANSwitcher`
+        weights_gen:(None, list, tuple)=None, # Weights for the generator and critic loss function
+        **kwargs
+    ):
```
Comment (warner-benjamin) [inline fastai/vision/gan.py:397]:
```suggestion
    def from_learners(cls,
        gen_learn:Learner, # A `Learner` object that has the generator
        crit_learn:Learner, # A `Learner` object that has the critic
        switcher:Callback=None, # Callback for switching between generator and critic training, defaults to `FixedGANSwitcher`
        weights_gen:(list, tuple)=None, # Weights for the generator and critic loss function
        **kwargs
    ):
```

**Review by warner-benjamin — COMMENTED** [review]
I don't think we need to add a None type hint for args with a default of None. I provided an example.

Remove the None type hints and it looks good.

File: fastai/vision/gan.py [hunk 6]
```diff
@@ -121,23 +160,33 @@ def forward(self, output, target):
 
 # Cell
 def accuracy_thresh_expand(y_pred, y_true, thresh=0.5, sigmoid=True):
```
Comment (kevinbird15) [inline fastai/vision/gan.py:162]:
```suggestion
def accuracy_thresh_expand(y_pred:Tensor, y_true:Tensor, thresh:float=0.5, sigmoid:bool=True):
```

May also want to docment this

File: fastai/vision/gan.py [hunk 7]
```diff
@@ -121,23 +160,33 @@ def forward(self, output, target):
```
Comment (kevinbird15) [inline fastai/vision/gan.py:160]: Add types to AdaptiveLoss

File: fastai/vision/gan.py [hunk 8]
```diff
@@ -121,23 +160,33 @@ def forward(self, output, target):
...
-    "Compute accuracy after expanding `y_true` to the size of `y_pred`."
+    "Compute thresholded accuracy after expanding `y_true` to the size of `y_pred`."
     if sigmoid: y_pred = y_pred.sigmoid()
     return ((y_pred>thresh).byte()==y_true[:,None].expand_as(y_pred).byte()).float().mean()
 
 # Cell
-def set_freeze_model(m, rg):
+def set_freeze_model(
+    m:nn.Module, # Model to freeze/unfreeze
+    rg:bool # `True` for freeze, `False` for unfreeze
```
Comment (kevinbird15) [inline fastai/vision/gan.py:170]:
```suggestion
    rg:bool # `Requires grad` argument. `True` for freeze
```

File: fastai/vision/gan.py [hunk 9]
```diff
@@ -259,7 +316,12 @@ class InvisibleTensor(TensorBase):
     def show(self, ctx=None, **kwargs): return ctx
```
Comment (kevinbird15) [inline fastai/vision/gan.py:316]: Maybe add doc string to explain `InvisibleTensor`.  What is Invisible about it?

File: fastai/vision/gan.py [hunk 10]
```diff
@@ -298,8 +364,21 @@ def _tk_diff(real_pred, fake_pred): return real_pred.mean() - fake_pred.mean()
...
+        critic:nn.Module, # Critic model
+        gen_loss_func:callable, # Generator loss function
+        crit_loss_func:callable, # Critic loss function
+        switcher:(Callback,None)=None, # Callback for switching between generator and critic training, defaults to `FixedGANSwitcher`
+        gen_first:bool=False, # Whether we start with generator training
+        switch_eval:bool=True, # Whether the model should be set to eval mode when calculating loss
+        show_img:bool=True, # Whether to show example generated images during training
+        clip:(None, float)=None, # How much to clip the weights
+        cbs:(Callback, None, list)=None, # Additional callbacks
+        metrics:(None, list, callable)=None, # Metrics
```
Comment (kevinbird15) [inline fastai/vision/gan.py:379]:
```suggestion
        metrics:(None, list, callable)=None, # `Metrics`
```
Hoping this will link to the fastai metrics doc, otherwise remove this comment

File: fastai/vision/gan.py [hunk 11]
```diff
@@ -309,14 +388,28 @@ def __init__(self, dls, generator, critic, gen_loss_func, crit_loss_func, switch
         super().__init__(dls, gan, loss_func=loss_func, cbs=cbs, metrics=metrics, **kwargs)
 
     @classmethod
-    def from_learners(cls, gen_learn, crit_learn, switcher=None, weights_gen=None, **kwargs):
+    def from_learners(cls,
+        gen_learn:Learner, # A `Learner` object that has the generator
```
Comment (kevinbird15) [inline fastai/vision/gan.py:392]:
```suggestion
        gen_learn:Learner, # A `Learner` object that contains the generator
```

File: fastai/vision/gan.py [hunk 12]
```diff
@@ -309,14 +388,28 @@ def __init__(self, dls, generator, critic, gen_loss_func, crit_loss_func, switch
...
+        crit_learn:Learner, # A `Learner` object that has the critic
```
Comment (kevinbird15) [inline fastai/vision/gan.py:393]:
```suggestion
        crit_learn:Learner, # A `Learner` object that contains the critic
```

File: fastai/vision/gan.py [hunk 13]
```diff
@@ -309,14 +388,28 @@ def __init__(self, dls, generator, critic, gen_loss_func, crit_loss_func, switch
...
+    def wgan(cls,
+        dls:DataLoaders, # DataLoaders object for GAN data
+        generator:nn.Module, # Generator model
+        critic:nn.Module, # Critic model
+        switcher:(Callback,None)=None, # Callback for switching between generator and critic training, defaults to `FixedGANSwitcher(n_crit=5, n_gen=1)`
+        clip:(None, float)=0.01, # How much to clip the weights
+        switch_eval:bool=False, # Whether the model should be set to eval mode when calculating loss
+        **kwargs
+    ):
+        "Create a WGAN from `dls`, `generator` and `critic`."
```
Comment (kevinbird15) [inline fastai/vision/gan.py:412]: What is a WGAN?  Is there a paper link or anything for this

**Review by kevinbird15 — CHANGES_REQUESTED** [review]

**Comment by warner-benjamin** [comment]
One other thought, maybe the model training should be behind a nbdev `#slow` tag so we are not replicating it every time tests are run on GH Actions. Would also be preferred not to check in the `UserWarning`s to the documentation

File: fastai/vision/gan.py (hunk 6 above)
Comment (tmabraham) [reply fastai/vision/gan.py:162]: thanks!

**Comment by tmabraham** [comment]
> One other thought, maybe the model training should be behind a nbdev `#slow` tag so we are not replicating it every time tests are run on GH Actions. Would also be preferred not to check in the `UserWarning`s to the documentation

There is an `all_slow` tag at the top of the notebook so that's not needed. IDK about the `UserWarning`s though. Technically, there is a [fastai PR](https://github.com/fastai/fastai/pull/3334) to fix this which we can resolve after this sprint.

**Comment by warner-benjamin** [comment]
Alright. Resolve the CI errors and then it looks good to me.

File: fastai/vision/gan.py (hunk 10 above)
Comment (tmabraham) [reply fastai/vision/gan.py]: It apparently doesn't

**Comment by tmabraham** [comment]
@warner-benjamin updated...

**Review by warner-benjamin — APPROVED** [review]
LGTM
